MINOR: extract jointly owned parts of BrokerServer and ControllerServer

cmccabe opened the pull request:
Extract jointly owned parts of BrokerServer and ControllerServer into JointServer. Shut down JointServer when the last component using it shuts down. (But make sure to stop the raft manager before closing the ControllerServer's sockets.)

cmccabe committed:
JointServer

hachikuji commented on core/src/main/scala/kafka/server/KafkaRaftServer.scala:
```diff
@@ -69,95 +64,56 @@ class KafkaRaftServer(
   }
 
   override def shutdown(): Unit = {
+    // In combined mode, we want to shut down the broker first, since it may take longer to complete.
```
We also need the controller to be available in order to process controlled shutdown.

cmccabe replied:
Good point. I'll rework this comment a bit.

hachikuji commented on core/src/test/scala/integration/kafka/server/QuorumTestHarness.scala:
```diff
@@ -99,25 +94,28 @@ class KRaftQuorumImplementation(
+      broker
+    } catch {
+      case e: Throwable => {
+        if (broker != null) CoreUtils.swallow(broker.shutdown(), log)
```
Shutdown `jointServer` here if `broker` is null? (Same thing for controller below.)

cmccabe replied:
Currently, if broker is null, then jointServer cannot have been started.

However, just to make it simple to understand what is going on, I'll add an unconditional call to `jointServer#stopForBroker` here

hachikuji commented on core/src/main/scala/kafka/server/JointServer.scala:
```diff
@@ -0,0 +1,237 @@
+   * The fault handler to use when metadata loading fails.
+   */
+  def metadataLoaderFaultHandler: FaultHandler = faultHandlerFactory.build("metadata loading",
+    config.processRoles.contains(ControllerRole),
```
nit: it's a little clearer if you provide the argument names, at least for the `fatal` arg. Same for other factories below.
```scala
fatal = config.processRoles.contains(ControllerRole)
```

cmccabe replied:
ok

hachikuji commented on core/src/main/scala/kafka/server/KafkaRaftServer.scala:
```diff
@@ -69,95 +64,56 @@ class KafkaRaftServer(
-    // to stick around until graceful shutdown of `RaftManager` can be completed.
-    raftManager.shutdown()
     controller.foreach(_.shutdown())
     CoreUtils.swallow(AppInfoParser.unregisterAppInfo(Server.MetricsPrefix, config.brokerId.toString, metrics), this)
```
Pre-existing issue, but I think we also need to close the `Metrics` instance.

cmccabe replied:
I will add this to `JointServer#stop`.

It was previously being done in `BrokerServer#shutdown`, which handles broker-only and combined mode. But I guess in the case of testing `ControllerServer`, we leaked this resource.

hachikuji commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -115,11 +114,33 @@ public void close() {
+                return fatalFaultHandler;
+            } else {
+                return nonFatalFaultHandler;
+
```
nit: remove newline

hachikuji commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -130,18 +151,56 @@ public Builder setConfigProp(String key, String value) {
+            props.put(KafkaConfig$.MODULE$.ControllerListenerNamesProp(),
+                    "CONTROLLER");
+            // Note: we can't accurately set controller.quorum.voters yet, since we don't
+            // yet know what ports each controller will pick.  Set it to a dummy string \
```
nit: extra `\` at the end of the line intended?

cmccabe replied:
that is weird. I think the IDE inserted it?

It does nothing here. I will remove it

hachikuji commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -159,53 +218,30 @@ public KafkaClusterTestKit build() throws Exception {
+                    String threadNamePrefix = (nodes.brokerNodes().containsKey(node.id())) ?
+                            String.format("colocated%d", node.id()) :
+                            String.format("controller%d", node.id());
+                    JointServer jointServer = new JointServer(createNodeConfig(node),
```
We end up doing a bunch of stuff that `KafkaRaftServer` is already doing. Seems like it would be simpler to use `KafkaRaftServer` directly and get rid of all this logic to manage the lower-level components. That would also get us more consistent shutdown behavior.

cmccabe replied:
It's not simple at all to use KafkaRaftServer in most of our tests. Let me give an example. If someone shuts down a broker in a test by calling BrokerServer#shutdown, and the broker was a standalone broker, you have to somehow shut down the associated KafkaRaftServer, the associated snapshot generator, and the associated metadata loader. And clear the associated dynamic metadata.

If you maintain KafkaRaftServer more or less the way it is, where it owns a BrokerServer, ControllerServer, and some other stuff, and those owned objects don't have any pointers back to it, this is not really possible. You would have to either rewrite the tests in terms of KafkaRaftServer, which is not really feasible in the time we have available, or just accept that BrokerServer#shutdown is not going to clean up everything. I don't think either course of action really works here.

In general we haven't tested combined mode very much, so we've been able to handwave some of this. Or just accept resource leaks in the tests. But to do it correctly, we should acknowledge that in combined mode there is some joint state. Hence, JointServer.

I think this PR greatly simplifies the test code (and will do so for the other test harnesses we have). We cannot have each test harness manually managing the joint state, it is just too much (and grows over time). This is a clean way to do that: standalone mode = your own JointServer, combined mode = shared JointServer.

cmccabe committed:
Address review comments

cmccabe committed:
Fix test memory leaks

cmccabe committed:
KafkaRaftClient#close: call releaseRetained

cmccabe committed:
Fix PlaintextAdminIntegrationTest.testCreateTopicsReturnsConfigs

Do not use colocation in PlaintextAdminIntegrationTest.testCreateTopicsReturnsConfigs since QuorumTestHarness does not really support colocation

Allow node configurations to be set for controllers

cmccabe committed:
Merge branch 'trunk' into jointserver

hachikuji commented on core/src/main/scala/kafka/server/BrokerServer.scala:
```diff
@@ -72,19 +67,14 @@ class BrokerSnapshotWriterBuilder(raftClient: RaftClient[ApiMessageAndVersion])
+  val config = jointServer.config
+  val time = jointServer.time
+  val metrics = jointServer.metrics
+  def raftManager: KafkaRaftManager[ApiMessageAndVersion] = jointServer.raftManager
```
nit: why not make this a `val` also?

cmccabe commented:
There were a few test flakes that don't look related. Will merge in trunk and get another test run, I guess.

hachikuji commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -214,61 +249,30 @@ metaProperties, config, new MetadataRecordSerde(), metadataPartition, KafkaRaftS
+                                JavaConverters.asScalaBuffer(Collections.<String>emptyList()).toSeq());
+                    } catch (Throwable e) {
+                        log.error("Error creating broker {}", node.id(), e);
+                        if (broker != null) broker.shutdown();
```
nit: similar comment as before. I think JointServer still has resources that need to be cleaned up even if it doesn't get started (e.g. `Metrics`). Same thing for the controller above.

cmccabe replied:
I think `Metrics` is the only one. It looks like its constructor starts a thread and so on.

But yes, we should shut down all that. Fixed.

hachikuji commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -2368,6 +2368,10 @@ public void close() {
         if (kafkaRaftMetrics != null) {
             kafkaRaftMetrics.close();
         }
+        if (memoryPool instanceof BatchMemoryPool) {
```
Perhaps we can let `MemoryPool` implement `Closeable`?

By the way, was this a leak? I think the only reference to the pool is in `KafkaRaftClient`, so does that mean we were leaking `KafkaRaftClient` references?

cmccabe replied:
I have been playing whack-a-mole with cases where we leak `KafkaRaftClient` instances in tests. Most of them seem to be related to the metrics closures dragging in giant objects, and the metrics not getting deleted properly. These are existing issues not caused by this PR.

This is a bit ugly but it stabilizes the build greatly so I think we should leave it in. I didn't want to make it a close() function since I didn't want to start thinking about dealing with a closed state in all of the memory pools.

Can we leave this for now and open a JIRA?

cmccabe commented:
Test failures not related
```
StickyAssignorTest > testLargeAssignmentAndGroupWithNonEqualSubscription() FAILED
CoordinatorTest > testTaskRequestWithOldStartMsGetsUpdated() FAILED
 MirrorConnectorsWithCustomForwardingAdminIntegrationTest > testReplicationIsCreatingTopicsUsingProvidedForwardingAdmin() FAILED
[DefaultTaskExecutorTest > shouldUnassignTaskWhenRequired() FAILED
 SmokeTestDriverIntegrationTest > shouldWorkWithRebalance() FAILED
```

cmccabe committed:
address review comments

hachikuji reviewed (approved):
Thanks, LGTM

cmccabe committed:
remove unused import

hachikuji commented:
The test failures in the latest build are showing an NPE:
```
java.lang.NullPointerException: Cannot invoke "kafka.raft.KafkaRaftManager.client()" because the return value of "kafka.server.BrokerServer.raftManager()" is null
	at kafka.server.BrokerServer.startup(BrokerServer.scala:304)
	at kafka.integration.KafkaServerTestHarness.$anonfun$createBrokers$1(KafkaServerTestHarness.scala:346)
	at kafka.integration.KafkaServerTestHarness.$anonfun$createBrokers$1$adapted(KafkaServerTestHarness.scala:342)
	at scala.collection.immutable.Vector.foreach(Vector.scala:1856)
	at kafka.integration.KafkaServerTestHarness.createBrokers(KafkaServerTestHarness.scala:342)
	at kafka.integration.KafkaServerTestHarness.setUp(KafkaServerTestHarness.scala:121)
	at kafka.api.IntegrationTestHarness.doSetup(IntegrationTestHarness.scala:123)
	at kafka.api.IntegrationTestHarness.setUp(IntegrationTestHarness.scala:103)
	at kafka.admin.AddPartitionsTest.setUp(AddPartitionsTest.scala:60)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
```

ijuma commented on core/src/main/scala/kafka/server/JointServer.scala:
```diff
@@ -0,0 +1,246 @@
+ * their reference. We opted to use two booleans here rather than a reference count in order to
+ * make debugging easier and reduce the chance of resource leaks.
+ */
+class JointServer(
```
Would `SharedServer` be a better name?

cmccabe replied:
yes, I suppose SharedServer might be better.

cmccabe commented:
> The test failures in the latest build are showing an NPE:

Yes. raftManager is initialized in `JointServer#start` so `BrokerServer#raftManager` needs to be a function, not a val.

cmccabe committed:
fix bug and rename jointserver -> sharedserver

- BrokerServer#raftManager must be a def, not val

- rename jointserver -> sharedserver

- fix one more failure case in a test harness where we needed to call stopForController to clean up the metrics object

hachikuji commented on core/src/main/scala/kafka/server/BrokerServer.scala:
```diff
@@ -66,14 +67,14 @@ class BrokerSnapshotWriterBuilder(raftClient: RaftClient[ApiMessageAndVersion])
+  val config = sharedServer.config
+  val time = sharedServer.time
+  val metrics = sharedServer.metrics
+  def raftManager: KafkaRaftManager[ApiMessageAndVersion] = sharedServer.raftManager
```
Can you add a comment here why this needs to be a `def`?

cmccabe committed:
support restarting SharedServer

cmccabe committed:
Merge branch 'trunk' into jointserver

cmccabe committed:
address review comment
